resources/assets/lib/beatmap-discussions/post.tsx

```diff
@@ -0,0 +1,457 @@
+                currentBeatmap: this.props.beatmap,
+                discussion: this.props.discussion,
+                user: this.props.user,
+              })}
+              user={this.props.user}
+            />
+          )}
+          {this.editing ? this.renderMessageEditor() : this.renderMessageViewer()}
+        </div>
+      </div>
+    );
+  }
+
+  private deleteHref(op: 'destroy' | 'restore') {
+    const [controller, key] = this.props.type === 'reply'
+      ? ['beatmapsets.discussions.posts', 'post']
+      : ['beatmapsets.discussions', 'discussion'];
+
+    return route(`${controller}.${op}`, { [key]: this.deleteModel.id });
+  }
+
+  @action
+  private readonly editCancel = () => {
+    this.editing = false;
+  };
+
+  @action
+  private readonly editStart = () => {
+    this.editorMinHeight = this.messageBodyRef.current != null
+      ? `${this.messageBodyRef.current.getBoundingClientRect().height + 50}px`
+      : '0';
+
+    this.editing = true;
+    this.message = this.props.post.message;
+  };
+
+  @action
+  private readonly handleEditorChange = () => {
+    this.canSave = this.reviewEditor.current?.canSave ?? false;
+  };
+
+  private readonly handleKeyDownCallback = (type: InputEventType) => {
+    if (type === InputEventType.Submit) {
+      this.updatePost();
+    }
+  };
+
+  private readonly handleMarkRead = () => {
+    $.publish('beatmapDiscussionPost:markRead', { id: this.props.post.id });
+  };
+
+  @action
+  private readonly handleTextareaChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
+    this.message = e.target.value;
+    this.canSave = validMessageLength(this.message, this.isTimeline);
+  };
+
+  private renderDeletedBy() {
+    if (this.deleteModel.deleted_at == null) return null;
+    const user = (
+      this.deleteModel.deleted_by_id != null
+        ? this.props.users[this.deleteModel.deleted_by_id]
+        : null
+    ) ?? deletedUser;
+
+    return (
+      <span className={`${bn}__info`}>
+        <StringWithComponent
+          mappings={{
+            delete_time: <TimeWithTooltip dateTime={this.deleteModel.deleted_at} relative />,
+            editor: (
+              <UserLink
+                className={`${bn}__info-user`}
+                user={user}
+              />
+            ),
+          }}
+          pattern={osu.trans('beatmaps.discussions.deleted')}
+        />
+      </span>
+    );
+  }
+
+  private renderEdited() {
+    if (this.props.lastEditor == null || this.props.post.updated_at === this.props.post.created_at) return null;
+
+    return (
+      <span className={`${bn}__info`}>
+        <StringWithComponent
+          mappings={{
+            editor: <UserLink className={`${bn}__info-user`} user={this.props.lastEditor} />,
+            update_time: <TimeWithTooltip dateTime={this.props.post.updated_at} relative />,
+          }}
+          pattern={osu.trans('beatmaps.discussions.edited')}
+        />
+      </span>
+    );
+  }
+
+  private renderKudosu() {
```

Comment:
Missing check `if @props.type == 'discussion' && @props.discussion.current_user_attributes?.can_moderate_kudosu`

Also there's another kudosu thing in this component (kudosu denied message) but this function name doesn't tell which one it is.